src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs

```diff
@@ -538,16 +538,47 @@ private static bool IsTypeLessVisibleThan(TypeSymbol type, Symbol sym, ref HashS
         /// traversal stops and that type is returned from this method. Otherwise if traversal
         /// completes without the predicate returning true for any type, this method returns null.
         /// </summary>
-        public static TypeSymbol VisitType<T>(this TypeSymbol type, Func<TypeSymbol, T, bool, bool> predicate, T arg, bool canDigThroughNullable =  false)
+        public static TypeSymbol VisitType<T>(
+            this TypeSymbol type,
+            Func<TypeSymbol, T, bool, bool> predicate,
+            T arg,
+            bool canDigThroughNullable = false)
+        {
+            return VisitType(
+                typeWithAnnotationsOpt: null,
+                typeOpt: type,
+                typeWithAnnotationsPredicateOpt: null,
+                typePredicateOpt: predicate,
+                arg,
+                canDigThroughNullable);
+        }
+
+        /// <summary>
+        /// Visit the given type and, in the case of compound types, visit all "sub type".
+        /// One of the predicates will be invoked at each type. If the type is a
+        /// TypeSymbolWithAnnotations, <paramref name="typeWithAnnotationsPredicateOpt"/>
+        /// will be invoked; otherwise <paramref name="typePredicateOpt"/> will be invoked.
+        /// If the corresponding predicate returns true for any type,
+        /// traversal stops and that type is returned from this method. Otherwise if traversal
+        /// completes without the predicate returning true for any type, this method returns null.
+        /// </summary>
+        public static TypeSymbol VisitType<T>(
```

Comment:
This signature, with two exclusive parameters, seems strange. Could we have one VisitType for TypeSymbol and one for TypeSymbolWithAnnotations?
Both would receive both predicates, and they would be mutually recursive (the overload for TSWA is a trivial wrapper for the other overload).

Reply:
One predicate is not sufficient in two cases: when the top-level type is a TypeSymbol rather than TypeSymbolWithAnnotations, and when walking up the containing types which are also TypeSymbols.

---
In reply to: [205834509](https://github.com/dotnet/roslyn/pull/28804#discussion_r205834509) [](ancestors = 205834509)

Reply:
I didn't mean bundle the predicates (I understand we need to keep two). I meant the `TypeSymbol typeOpt` and `TSWA ...` parameters.

Reply:
I agree, it's strange to have two mutually exclusive parameters. But it does allow writing specific helper methods that encapsulate the call to `VisitType` but that also take either a `TypeSymbol` or a `TypeSymbolWithAnnotations` (see `TypeSymbolWithAnnotations.ContainsNullableReferenceTypes for instance).

---
In reply to: [205853384](https://github.com/dotnet/roslyn/pull/28804#discussion_r205853384) [](ancestors = 205853384)